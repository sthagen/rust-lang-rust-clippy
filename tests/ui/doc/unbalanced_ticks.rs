//! This file tests for the `DOC_MARKDOWN` lint, specifically cases
//! where ticks are unbalanced (see issue #6753).
//@no-rustfix
#![allow(dead_code)]
#![warn(clippy::doc_markdown)]

/// This is a doc comment with `unbalanced_tick marks and several words that
//~^ empty_line_after_doc_comments

/// should be `encompassed_by` tick marks because they `contain_underscores`.
/// Because of the initial `unbalanced_tick` pair, the error message is
/// very `confusing_and_misleading`.
fn main() {}

/// This paragraph has `unbalanced_tick marks and should stop_linting.
//~^ empty_line_after_doc_comments

///
/// This paragraph is fine and should_be linted normally.

///
/// Double unbalanced backtick from ``here to here` should lint.

///
/// Double balanced back ticks ``start end`` is fine.
fn multiple_paragraphs() {}

/// ```
/// // Unbalanced tick mark in code block shouldn't warn:
/// `
/// ```
fn in_code_block() {}

/// # `Fine`
///
/// ## not_fine
//~^ empty_line_after_doc_comments

///
/// ### `unbalanced

///
/// - This `item has unbalanced tick marks

/// - This item needs backticks_here

fn other_markdown() {}

#[rustfmt::skip]
/// - ```rust
///   /// `lol`
///   pub struct Struct;
///   ```
fn issue_7421() {}

/// `
//~^ empty_line_after_doc_comments

fn escape_0() {}

/// Escaped \` backticks don't count.
fn escape_1() {}

/// Escaped \` \` backticks don't count.
fn escape_2() {}

/// Escaped \` ` backticks don't count, but unescaped backticks do.
//~^ empty_line_after_doc_comments

fn escape_3() {}

/// Backslashes ` \` within code blocks don't count.
fn escape_4() {}
