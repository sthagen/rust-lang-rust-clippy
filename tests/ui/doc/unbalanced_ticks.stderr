error: empty line after doc comment
  --> tests/ui/doc/unbalanced_ticks.rs:7:1
   |
LL | / /// This is a doc comment with `unbalanced_tick marks and several words that
LL | |
LL | |
   | |_^
...
LL |   fn main() {}
   |   --------- the comment documents this function
   |
   = note: `-D clippy::empty-line-after-doc-comments` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::empty_line_after_doc_comments)]`
   = help: if the empty line is unintentional remove it
help: if the doc comment should not document `main` comment it out
   |
LL | // /// This is a doc comment with `unbalanced_tick marks and several words that
   | ++

error: empty lines after doc comment
  --> tests/ui/doc/unbalanced_ticks.rs:15:1
   |
LL | / /// This paragraph has `unbalanced_tick marks and should stop_linting.
LL | |
LL | |
LL | | ///
...  |
LL | | /// Double unbalanced backtick from ``here to here` should lint.
LL | |
   | |_^
...
LL |   fn multiple_paragraphs() {}
   |   ------------------------ the comment documents this function
   |
   = help: if the empty lines are unintentional remove them
help: if the doc comment should not document `multiple_paragraphs` comment it out
   |
LL ~ // /// This paragraph has `unbalanced_tick marks and should stop_linting.
LL |
LL |
LL ~ // ///
LL ~ // /// This paragraph is fine and should_be linted normally.
LL |
LL ~ // ///
LL ~ // /// Double unbalanced backtick from ``here to here` should lint.
   |

error: empty lines after doc comment
  --> tests/ui/doc/unbalanced_ticks.rs:36:1
   |
LL | / /// ## not_fine
LL | |
LL | |
LL | | ///
...  |
LL | | /// - This item needs backticks_here
LL | |
   | |_^
LL |   fn other_markdown() {}
   |   ------------------- the comment documents this function
   |
   = help: if the empty lines are unintentional remove them
help: if the doc comment should not document `other_markdown` comment it out
   |
LL ~ // /// # `Fine`
LL ~ // ///
LL ~ // /// ## not_fine
LL |
LL |
LL ~ // ///
LL ~ // /// ### `unbalanced
LL |
LL ~ // ///
LL ~ // /// - This `item has unbalanced tick marks
LL |
LL ~ // /// - This item needs backticks_here
   |

error: empty line after doc comment
  --> tests/ui/doc/unbalanced_ticks.rs:56:1
   |
LL | / /// `
LL | |
LL | |
   | |_^
LL |   fn escape_0() {}
   |   ------------- the comment documents this function
   |
   = help: if the empty line is unintentional remove it
help: if the doc comment should not document `escape_0` comment it out
   |
LL | // /// `
   | ++

error: empty line after doc comment
  --> tests/ui/doc/unbalanced_ticks.rs:67:1
   |
LL | / /// Escaped \` ` backticks don't count, but unescaped backticks do.
LL | |
LL | |
   | |_^
LL |   fn escape_3() {}
   |   ------------- the comment documents this function
   |
   = help: if the empty line is unintentional remove it
help: if the doc comment should not document `escape_3` comment it out
   |
LL | // /// Escaped \` ` backticks don't count, but unescaped backticks do.
   | ++

error: aborting due to 5 previous errors

